File: CA_Server/Server/Game/Object/GameObject.cs
﻿using Google.Protobuf.Protocol;

namespace Server.Game
{
    public class GameObject
    {
        public GameObjectType ObjectType { get; protected set; } = GameObjectType.None;
        public int Id
        {
            get { return Info.ObjectId; }
            set { Info.ObjectId = value; }
        }
        public GameRoom Room { get; set; }
        public ObjectInfo Info { get; set; } = new ObjectInfo();
        public PositionInfo PosInfo { get; private set; } = new PositionInfo();

        public GameObject()
        {
            Info.PosInfo = PosInfo;
        }

        public Vector2Int CellPos
        {
            get { return new Vector2Int(PosInfo.PosX, PosInfo.PosY); }
            set
            {
                PosInfo.PosX = value.x;
                PosInfo.PosY = value.y;
            }
        }

        public Vector2Int GetFrontCellPos()
        {
            return GetFrontCellPos(PosInfo.MoveDir);
        }

        public Vector2Int GetFrontCellPos(MoveDir dir, int distance = 1)
        {
            Vector2Int cellPos = CellPos;

            switch (dir)
            {
                case MoveDir.Up:
                    cellPos += Vector2Int.up * distance;
                    break;
                case MoveDir.Down:
                    cellPos += Vector2Int.down * distance;
                    break;
                case MoveDir.Left:
                    cellPos += Vector2Int.left * distance;
                    break;
                case MoveDir.Right:
                    cellPos += Vector2Int.right * distance;
                    break;
            }

            return cellPos;
        }
    }
}

File: CA_CppServer/GameServer/GameObject.cpp
#include "pch.h"
#include "GameObject.h"

void GameObject::SetCellPos(int32 x, int32 y)
{
    Protocol::PositionInfo* posInfo = info.mutable_posinfo();
    posInfo->set_posx(x);
    posInfo->set_posy(y);
}

Vector2Int GameObject::GetFrontCellPos(Protocol::MoveDir dir, int32 distance)
{
    Vector2Int cellPos = GetCellPos();

    switch (dir)
    {
    case Protocol::MoveDir::UP:
        cellPos = cellPos + Vector2Int::up() * distance;
        break;
    case Protocol::MoveDir::DOWN:
        cellPos = cellPos + Vector2Int::down() * distance;
        break;
    case Protocol::MoveDir::LEFT:
        cellPos = cellPos + Vector2Int::left() * distance;
        break;
    case Protocol::MoveDir::RIGHT:
        cellPos = cellPos + Vector2Int::right() * distance;
        break;
    }

    return cellPos;

}